src/graphapi/operationgraph_descriptor.cpp

```diff
@@ -0,0 +1,237 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/errors.hpp>
+#include <miopen/graphapi/operationgraph_descriptor.hpp>
+
+#include <algorithm>
+
+namespace miopen {
+
+namespace graphapi {
+
+OperationGraph::OperationGraph(miopenHandle_t handle, const OpGraph& opGraph) : mHandle(handle)
```

Comment:
Some minor issues:
-- `deref(miopenHandle_t` so that this class stores `Handle` pointer
-- A bit of a naming problem where we now have `OpGraph` and `OperationGraph`. We could call this class `OpGraphDesc` since it deals with descriptor related stuff. I'm open to other ideas as my suggestion is not that great.

Reply:
I think then we will have to add a suffix to all existing classes related to descriptors: Tensor, Convolution, Pointwise, Rng, VariantPack...

OpGraph seems to be used internally only and it doesn't have a corresponding descriptor. We can move it into an internal namespace.

Another option is to merge these 2 classes into one.

Reply:
And another line of classes are BackendTensorDescriptor, BackendVariantPackDescriptor already has -descriptor suffix, this will be confusing with OperationGraphDesc. Moreover I thought we might remove Backend- prefix in future as this API is no longer called Backend API.

Also I didn't get you about deref, do you suggest to deref `handle`?

Reply:
For me this `OperationGraph` class looks more like `EngineList`, may be call it this way?

The disadvantage of `EngineList` is that it is too far from its enum name `MIOPEN_BACKEND_OPERATIONGRAPH_DESCRIPTOR`.

CuDNN frontend API has such a class and they called it OperationGraph: https://github.com/NVIDIA/cudnn-frontend/blob/1b0b5eac540b7f8fd19b18f1e6b8427c95503348/include/cudnn_frontend_OperationGraph.h#L265-L266

Reply:
I think the most sensible thing here is to combine `OpGraph` and `OperationGraph` classes and their builders. A couple of points:
1. I do not think of `OpGraph` as an internal class. It's the operation graph for C++ users
2. CUDNN made the decision to put the `handle` in the graph. It does not make sense to me (they must have a reason). To me graph is just a description of operations and their dependencies. It shouldn't even be returning a list of `Engines`, but since we're following them blindly, we do what the API demands.

Reply:
Agree

Reply:
done